[1108] MP Rest Client 2.0 - First Look [staging] (#1116)
OpenLiberty/blogs: posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc

. <<sse-support,Support for Server Sent Events (SSEs)>>

== Getting started

The MP Rest Client 2.0 implementation is new in Open Liberty 21.0.0.3 - so make sure you are using the latest
version of Liberty. Next, you will need to add the following link:{url-prefix}/guides/maven-intro.html[Maven] dependency
to your pom.xml:

[source,xml]
----
<dependency>
    <groupId>org.eclipse.microprofile.rest.client</groupId>
    <artifactId>microprofile-rest-client-api</artifactId>
    <version>2.0</version>
    <scope>provided</scope>
</dependency>
----

or, if you use link:{url-prefix}/guides/gradle-intro.html[Gradle]:
[source,gradle]
----
dependencies {
    mpRestClient group: 'org.eclipse.microprofile.rest.client', name: 'microprofile-rest-client-api', version: '2.0'
}
----

Also, make sure to configure your Liberty server with the `mpRestClient-2.0` feature in the server.xml:
[source,xml]
----
<server>
  <featureManager>
    <feature>mpRestClient-2.0</feature>
    <!-- ... -->
  </featureManager>
  <!-- ... -->
</server>
----

That's it! Now that we've got our development and deployment environments set up, it's time to code!

[#query-param-style]

> == Use QueryParamStyle for specifying how collections of query parameters should be formatted

Reviewer: ```suggestion
== Using QueryParamStyle to specify how collections of query parameters should be formatted
```

Author: Done

Rest Client interfaces can specify query parameters using the `@QueryParam("paramName")` annotation, but often a server
will require that multi-valued query parameters must be formatted in a certain way. For example, suppose we have a
client interface like:
[source,java]
----
@RegisterRestClient
public interface MyMultiValuedQueryParamClient {

    @GET
    String multiValues(@QueryParam("myParam") List<String> values);
}
----
By default, if a caller invokes the the client with `multiValues(Arrays.asList("a", "b", "c"))` the MP Rest Client will
produce an HTTP request with multiple key/value pairs - something like: `?myParam=a&myParam=b&myParam=c`

Most servers will handle that just fine, but other servers will require that the HTTP request be a single key with a 
comma-separated list of values, like: `?myParam=a,b,c`

> Still other servers will require array-like syntax, such as: `?myParam[]=a&myParam[]=b&myParam[]=c`

Reviewer 2: ```suggestion
Other servers will still require array-like syntax, such as: `?myParam[]=a&myParam[]=b&myParam[]=c`
```

Author: Sounds good.

In order to support those other server types, you can now use the `QueryParamStyle` enum when building the client
instance - for example:

[source,java]
----
MyMultiValuedQueryParamClient client = RestClientBuilder.newBuilder()
                                                        .queryParamStyle(QueryParamStyle.COMMA_SEPARATED)
                                                        //...
                                                        .build(MyMultiValuedQueryParamClient.class);
----

Alternatively the query parameter style can be declared through MP Config using a property like:

> `com.mypkg.MyMultiValuedQueryParamClient/mp-rest/queryParamStyle=ARRAY_PAIRS`

Reviewer: @andymc12 should we have a short table that lists the three options and gives an example of each. As it is you kinda have to infer which one is which?

> `com.mypkg.MyMultiValuedQueryParamClient/mp-rest/queryParamStyle=ARRAY_PAIRS`

Reviewer: Also, `MyMultiValuedQueryParamClient` is a rather long class name. The other sections below just use `MyClient`, that might be better here too?

Author: Sure - switched to `MyClient`

[#proxy-server]
== Proxy server support

> Many RESTful endpoints might require the use of a proxy server to access. MicroProfile Rest Client 2.0 makes it easier

Reviewer 2: I think we should either have the first "Microprofile" expanded and then all others as "MP" or just stick to "MicroProfile" right now sometimes it's "MP" and sometimes it's "MicroProfile"

Author: Agreed - I will change these to be "MP" after the first time it is referenced in the text.  I may leave it as "MicroProfile" in the closing summary though - let me know if you have a strong opinion against that.

Reviewer 2: Really good idea!

and more portable to specify a proxy server with the new `proxyAddress(host, port)` method on the `RestClientBuilder`
class. For example, suppose you need to access an endpoint via a proxy server at myproxy.xyz.com on port 1080. You could
build your rest client instance with code like:

[source,java]
----
MyClient client = RestClientBuilder.newBuilder()
                                   .proxyAddress(myproxy.xyz.com, 1080)
                                   //...
                                   .build(MyClient.class);
----

Alternatively, you can specify the proxy address via MP Config with a property like:

`com.mypkg.MyClient/mp-rest/proxyAddress=myproxy.xyz.com:1080`

Note that this approach to setting the proxy server host and port should be preferred to using vendor-specific

> properties like `com.ibm.ws.jaxrs.client.proxy.host` and `com.ibm.ws.jaxrs.client.proxy.port`, though these properties

Reviewer: ```suggestion
properties such as `com.ibm.ws.jaxrs.client.proxy.host` and `com.ibm.ws.jaxrs.client.proxy.port`, though these properties
```

will still work. For proxy authentication, you can still use the `com.ibm.ws.jaxrs.client.proxy.username` and

> `com.ibm.ws.jaxrs.client.proxy.password` properties like:

Reviewer: ```suggestion
`com.ibm.ws.jaxrs.client.proxy.password` properties. For example:
```

[source,java]
----
MyClient client = RestClientBuilder.newBuilder()
                                   .proxyAddress(myproxy.xyz.com, 1080)
                                   .property("com.ibm.ws.jaxrs.client.proxy.username", "andymc12")
                                   .property("com.ibm.ws.jaxrs.client.proxy.password", "12345") //same as my luggage! :)
                                   //...
                                   .build(MyClient.class);
----

[#auto-redirects]
== Automatically following redirects

> If a RESTful resource has relocated, often the HTTP response code will be in the 300 range and will indicate the new

Reviewer: ```suggestion
If a RESTful resource has been relocated, often the HTTP response code will be in the 300 range and will indicate the new
```

location. Rather than handling the 3XX response and manually issuing a new request, MP Rest Client 2.0 allows rest
client instances to automatically follow redirects. This can be configured programmatically while building the client
instance or via MP Config. Here is an example of configuring auto-redirect via the `RestClientBuilder` API:

[source,java]
----
MyClient client = RestClientBuilder.newBuilder()
                                   .followRedirects(true)
                                   //...
                                   .build(MyClient.class);
----

And here is how you would configure it via MP Config:

`com.mypkg.MyClient/mp-rest/followRedirects=true`

[#sse-support]
== Support for Server Sent Events (SSEs)

link:https://openliberty.io/guides/reactive-messaging-sse.html[Server Sent Events] are part of the HTML 5 spec and

> enables a server to push data to a client asynchronously via events - all over HTTP. The JAX-RS 2.1 spec enabled SSE

Reviewer 2: Is `JAX-RS 2.1` not supposed to be in backticks?

Author: In this case, I don't think it should be in backticks.  I'm referring to the specification/project, not the Liberty feature (which would go in backticks, like `jaxrs-2.1`).

support for both the client and server. Now you can consume SSE events from the type-safe MP Rest Client.

The MP Rest Client specification uses the link:http://www.reactive-streams.org/[Reactive Streams] APIs to consume
events. A client interface capable of consuming SSEs would look something like this:

[source,java]
----
@RegisterRestClient
public interface SseClient {

    @GET
    @Path("/path/sse")
    @Produces(MediaType.SERVER_SENT_EVENTS)
    Publisher<String> getStrings();

    @GET
    @Path("/path/sse2")
    @Produces(MediaType.SERVER_SENT_EVENTS)
    Publisher<InboundSseEvent> getEvents();
}
----

First, the method (or interface) must be annotated with `@Produces(MediaType.SERVER_SENT_EVENTS)` to indicate that it
expects the server to produce SSEs. Next, the method's return type must be `org.reactivestreams.Publisher`. The generic

> type can be `javax.ws.rs.sse.InboundSseEvent` from JAX-RS or it can be a primitive or complex object - the latter only
> works if (1) the server only sends one type of events (e.g. only sends `WeatherEvents` - then `Publisher<WeatherEvent>`
> would be applicable) and (2) if you have an entity provider that is capable of converting the events into the generic
> type (e.g. `MessageBodyReader<WeatherEvent>`). For most cases, if the remote server sends events using JSON, you can
> enable the `jsonb-1.0` feature in your Liberty server which will automatically register a JSON-B-based entity provider.

Reviewer: ```suggestion
type can be `javax.ws.rs.sse.InboundSseEvent` (from JAX-RS), a primitive, a String or a complex type. Complex types may only be used if;
 1. the server only sends one type of event (e.g. only sends `WeatherEvents` - then `Publisher<WeatherEvent>` would be applicable)
_and_
 2. there is a registered entity provider capable of converting the events (e.g. `MessageBodyReader<WeatherEvent>`).
In most cases, if the remote server sends events using JSON, you can
enable the `jsonb-1.0` feature in your Liberty server which will automatically register a JSON-B-based entity provider.
```

Once you've invoked one of these methods, you should be able to register one or more `Subscriber` instances to the
`Publisher`. Each subscriber will be notified on receipt of a new event or if the connection to the server has been
closed.

== Summary

> MicroProfile Rest Client 2.0 has some powerful new features useful for building cloud native applications. There are
> still more features not covered here, but have been in Open Liberty's previous Rest Client support, including using CDI-
> enabled providers and supporting `ContextResolver` to customize the built-in JSON-B provider. These features and more
> are available on the link:https://github.com/eclipse/microprofile-rest-client/releases/tag/2.0[MP Rest Client 2.0 release page].

Reviewer: Does "JSON-B ContextResolver Support" not deserve it's own section? If not then it seems a little confusing to mention it here.
```suggestion
MicroProfile Rest Client 2.0 has some powerful new features useful for building cloud native applications. You can read more about these updates on the link:https://github.com/eclipse/microprofile-rest-client/releases/tag/2.0[MP Rest Client 2.0 release page].
```

---
layout: post
title: "MicroProfile Rest Client 2.0 - First Look"
categories: blog
author_picture: https://avatars3.githubusercontent.com/andymc12
author_github: https://github.com/andymc12
seo-title: MicroProfile Rest Client 2.0 - First Look - OpenLiberty.io
seo-description: This post describes the new features available in MicroProfile Rest Client 2.0 and how to use
them.
blog_description: "This post describes the new features available in MicroProfile Rest Client 2.0 and how to use
them."
---
= MicroProfile Rest Client 2.0 - First Look
Andy McCright <https://github.com/andymc12>

The latest release of MicroProfile's type-safe REST client has a lot of new and exciting features. In this post, we'll
take a look at some of the new features and how you can use them in Open Liberty, including:

> . <<query-param-style,Using QueryParamStyle to specify how collections of query parameters should be formatted>>

Reviewer 3: I think this list should be unordered bullets rather than ordered steps, since it doesn't list sequential, task based info

Author: Agreed - will fix

. <<sse-support,Support for Server Sent Events (SSEs)>>

== Getting started

The MP Rest Client 2.0 implementation is new in Open Liberty 21.0.0.3 - so make sure you are using the latest
version of Liberty. Next, you will need to add the following link:{url-prefix}/guides/maven-intro.html[Maven] dependency
to your pom.xml:

[source,xml]
----
<dependency>
    <groupId>org.eclipse.microprofile.rest.client</groupId>
    <artifactId>microprofile-rest-client-api</artifactId>
    <version>2.0</version>
    <scope>provided</scope>
</dependency>
----

or, if you use link:{url-prefix}/guides/gradle-intro.html[Gradle]:
[source,gradle]
----
dependencies {
    mpRestClient group: 'org.eclipse.microprofile.rest.client', name: 'microprofile-rest-client-api', version: '2.0'
}
----

Also, make sure to configure your Liberty server with the `mpRestClient-2.0` feature in the server.xml:
[source,xml]
----
<server>
  <featureManager>
    <feature>mpRestClient-2.0</feature>
    <!-- ... -->
  </featureManager>
  <!-- ... -->
</server>
----

That's it! Now that we've got our development and deployment environments set up, it's time to code!

[#query-param-style]

> == Using QueryParamStyle to specify how collections of query parameters should be formatted

Reviewer 3: suggested edit:

Using QueryParamStyle to specify how to format collections of query parameters 

If you agree, reminder to change the heading in the introduction list as well

Author: I think this should remain as it is.  I don't like the passive voice or the extra words either - but I think it better implies that the implementation (Liberty) handles the formatting - the user just tells us how to format it with the `QueryParamStyle`.

Rest Client interfaces can specify query parameters using the `@QueryParam("paramName")` annotation, but often a server
will require that multi-valued query parameters must be formatted in a certain way. For example, suppose we have a
client interface like:
[source,java]
----
@RegisterRestClient
public interface MyClient {
    @GET
    String multiValues(@QueryParam("myParam") List<String> values);
}
----

> By default, if a caller invokes the the client with `multiValues(Arrays.asList("a", "b", "c"))` the MP Rest Client will

Reviewer 3: remove extra "the" 

if a caller invokes the the client

and add a period after 

`?myParam=a&myParam=b&myParam=c`

Author: Good catch! will fix

produce an HTTP request with multiple key/value pairs - something like: `?myParam=a&myParam=b&myParam=c`

> Most servers will handle that just fine, but other servers will require that the HTTP request be a single key with a 

Reviewer 3: suggested edit

Most servers will handle that just fine, but other servers will require that the HTTP request be a single key ...

--->

Although most servers will handle that just fine, some servers might require the HTTP request to be a single key...

comma-separated list of values, like: `?myParam=a,b,c`

Other servers will require array-like syntax, such as: `?myParam[]=a&myParam[]=b&myParam[]=c`

In order to support those other server types, you can now use the `QueryParamStyle` enum when building the client
instance - for example:

[source,java]
----
MyClient client = RestClientBuilder.newBuilder()
                                   .queryParamStyle(QueryParamStyle.COMMA_SEPARATED)
                                   //...
                                   .build(MyClient.class);
----

> Alternatively the query parameter style can be declared through MP Config using a property like:

Reviewer 3: Alternatively, you can declare the query parameter style through MP Config by using a property like:

Reviewer 3: Also, is it helpful to link to MP Config guide on first mention of MP Config?
In case the reader is not familiar...

https://www.openliberty.io/guides/microprofile-config.html or https://www.openliberty.io/guides/microprofile-config-intro.html

Author: Good idea - will do!

`com.mypkg.MyClient/mp-rest/queryParamStyle=ARRAY_PAIRS`

.QueryParamStyle Options

> |=======================

Reviewer 3: I think this table needs an intro sentence -does it apply specifically to MP config or just generally to Rest Client?

Something like

"The following table lists the enum values that are available for xxxxx and provides an example oof the corresponding output for each value"

|Enum Value            | Output Example
|MULTI_PAIRS (default) | `?myParam=a&myParam=b&myParam=c`
|COMMA_SEPARATED       | `?myParam=a,b,c`
|ARRAY_PAIRS           | `?myParam[]=a&myParam[]=b&myParam[]=c`
|=======================

[#proxy-server]
== Proxy server support

> Many RESTful endpoints might require the use of a proxy server to access. MP Rest Client 2.0 makes it easier

Reviewer 3: suggested edit
Many RESTful endpoints might require the use of a proxy server to access.

--->

You might need to use a proxy server to access some RESTful endpoints.

and more portable to specify a proxy server with the new `proxyAddress(host, port)` method on the `RestClientBuilder`
class. For example, suppose you need to access an endpoint via a proxy server at myproxy.xyz.com on port 1080. You could

> build your rest client instance with code like:

Reviewer 3: should these values be monospaced?

suppose you need to access an endpoint via a proxy server at `myproxy.xyz.com` on port `1080`.

[source,java]
----
MyClient client = RestClientBuilder.newBuilder()
                                   .proxyAddress(myproxy.xyz.com, 1080)
                                   //...
                                   .build(MyClient.class);
----

Alternatively, you can specify the proxy address via MP Config with a property like:

`com.mypkg.MyClient/mp-rest/proxyAddress=myproxy.xyz.com:1080`

> Note that this approach to setting the proxy server host and port should be preferred to using vendor-specific

Reviewer 3: should be preferred
--->
is preferred.

Also, is it possible to concisely state why it's better/preferred to use MP config than the vendor specific properties?

properties such as `com.ibm.ws.jaxrs.client.proxy.host` and `com.ibm.ws.jaxrs.client.proxy.port`, though these
properties will still work. For proxy authentication, you can still use the `com.ibm.ws.jaxrs.client.proxy.username` and
`com.ibm.ws.jaxrs.client.proxy.password` properties. For example:
[source,java]
----
MyClient client = RestClientBuilder.newBuilder()
                                   .proxyAddress(myproxy.xyz.com, 1080)
                                   .property("com.ibm.ws.jaxrs.client.proxy.username", "andymc12")
                                   .property("com.ibm.ws.jaxrs.client.proxy.password", "12345") //same as my luggage! :)
                                   //...
                                   .build(MyClient.class);
----

[#auto-redirects]
== Automatically following redirects

If a RESTful resource has been relocated, often the HTTP response code will be in the 300 range and will indicate the
new location. Rather than handling the 3XX response and manually issuing a new request, MP Rest Client 2.0 allows rest
client instances to automatically follow redirects. This can be configured programmatically while building the client

> instance or via MP Config. Here is an example of configuring auto-redirect via the `RestClientBuilder` API:

Reviewer 3: This can be configured programmatically while building the client instance or via MP Config.

--->

You can configure a client to automatically follow redirects either programmatically, when you build the client instance, or via MP Config.

[source,java]
----
MyClient client = RestClientBuilder.newBuilder()
                                   .followRedirects(true)
                                   //...
                                   .build(MyClient.class);
----

And here is how you would configure it via MP Config:

`com.mypkg.MyClient/mp-rest/followRedirects=true`

[#sse-support]
== Support for Server Sent Events (SSEs)

link:https://openliberty.io/guides/reactive-messaging-sse.html[Server Sent Events] are part of the HTML 5 spec and

> enables a server to push data to a client asynchronously via events - all over HTTP. The JAX-RS 2.1 spec enabled SSE

Reviewer 3: [Server Sent Events] are part of the HTML 5 spec and enables a server to push data to a client asynchronously via events - all over HTTP.

--->

[Server Sent Events], part of the HTML 5 spec, enable a server to push data to a client asynchronously via events,  over HTTP.

support for both the client and server. Now you can consume SSE events from the type-safe MP Rest Client.

The MP Rest Client specification uses the link:http://www.reactive-streams.org/[Reactive Streams] APIs to consume

> events. A client interface capable of consuming SSEs would look something like this:

Reviewer 3: A client interface capable of consuming SSEs would look something like this:

--->

A client interface capable of consuming SSEs looks something like this:

[source,java]
----
@RegisterRestClient
public interface SseClient {

    @GET
    @Path("/path/sse")
    @Produces(MediaType.SERVER_SENT_EVENTS)
    Publisher<String> getStrings();

    @GET
    @Path("/path/sse2")
    @Produces(MediaType.SERVER_SENT_EVENTS)
    Publisher<InboundSseEvent> getEvents();
}
----

First, the method (or interface) must be annotated with `@Produces(MediaType.SERVER_SENT_EVENTS)` to indicate that it
expects the server to produce SSEs. Next, the method's return type must be `org.reactivestreams.Publisher`. The generic

> type can be `javax.ws.rs.sse.InboundSseEvent` (from JAX-RS), a primitive, a String or a complex type. Complex types may only be used if;

Reviewer 3: use serial comma:

The generic type can be `javax.ws.rs.sse.InboundSseEvent` (from JAX-RS), a primitive, a String, or a complex type.
 
Also, avoid "may"

Complex types can only be used if....

1) the server only sends one type of event (e.g. only sends `WeatherEvents` - then `Publisher<WeatherEvent>` would be applicable)

_and_

2) there is a registered entity provider capable of converting the events (e.g. `MessageBodyReader<WeatherEvent>`).

> In most cases, if the remote server sends events using JSON, you can enable the `jsonb-1.0` feature in your Liberty

Reviewer 3: add comma:

you can enable the jsonb-1.0 feature in your Liberty server, which automatically registers a JSON-B-based entity provider.

server which will automatically register a JSON-B-based entity provider.

> Once you've invoked one of these methods, you should be able to register one or more `Subscriber` instances to the

Reviewer 3: avoid past perfect and subjunctive:

Once you invoke one of these methods, you can register one or more `Subscriber` instances...

`Publisher`. Each subscriber will be notified on receipt of a new event or if the connection to the server has been
closed.

== Summary

> MicroProfile Rest Client 2.0 has some powerful new features useful for building cloud native applications. You can read

Reviewer 3: MicroProfile Rest Client 2.0 has some powerful new features useful for building cloud native applications.
--->
MicroProfile Rest Client 2.0 has some powerful new features that are useful for building cloud native applications.
